feat(gitlab): allow override server version (#11416)
renovatebot/renovate: lib/platform/gitlab/index.md

## Features awaiting implementation

- The `automergeStrategy` configuration option has not been implemented for this platform, and all values behave as if the value `auto` was used. Renovate will accept the Merge Request without further configuration, and respect the strategy defined in the Merge Request, and this cannot be overridden yet

> ## Server version dependend features

Reviewer: ```suggestion
## Server version dependent features
```

> We use the GitLab [version api](https://docs.gitlab.com/ee/api/version.html) to fetch the server version.

Reviewer: ```suggestion
We use the GitLab [version API](https://docs.gitlab.com/ee/api/version.html) to fetch the server version.
```

> This can be overridden by the experimental feature flag [`RENOVATE_X_SERVER_VERSION`](https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_server_version).

Reviewer: ```suggestion
You can use the experimental feature flag [`RENOVATE_X_SERVER_VERSION`](https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_server_version) to set a specific server version.
By setting the server version yourself, you save a API call that fetches the server version.
```

Would this be better?

- Use `Draft:` MR prefix instead of `WIP:` prefix since `v13.2.0`

> - Do not truncate markdown body to 25K since `v13.4.0`

Reviewer: ```suggestion
- Do not truncate Markdown body to 25K since `v13.4.0`
```

What is `25K`?  Do we mean kilobytes? Or do we mean something else?

> - Do not truncate Markdown body to 25K since `v13.4.0`

Reviewer: Okay! Thanks for explaining!

You can apply the suggestion yourself. 😉

> - Allow configure reviewers since `v13.9.0`

Reviewer: This list is a bit hard to understand, what do those versions mean? I think we mean to say that "certain features are only available after certain GitLab" versions? But this is not really explained properly in the current wording.

I think we should have a new sentence here that explains what those versions mean.

Author: Yes, that are the gitlab server versions, when renovate will change the behavior

Reviewer: I'm also not sure, because this is complicated stuff... 😄 

Maybe @rarkins can propose a draft text that we can improve on together?

## Features awaiting implementation

- The `automergeStrategy` configuration option has not been implemented for this platform, and all values behave as if the value `auto` was used. Renovate will accept the Merge Request without further configuration, and respect the strategy defined in the Merge Request, and this cannot be overridden yet

## Server version dependent features

We use the GitLab [version API](https://docs.gitlab.com/ee/api/version.html) to fetch the server version.
You can use the experimental feature flag [`RENOVATE_X_SERVER_VERSION`](https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_server_version) to set a specific server version.

> By setting the server version yourself, you save a API call that fetches the server version.

Author: 🤔  This needs to be set if you like to use `CI_JOB_TOKEN` because it doesn't have access right to the version api.

Author: It's best to use `CI_SERVER_VERSION` to set this in gitlab pipelines

Reviewer: > 🤔 This needs to be set if you like to use `CI_JOB_TOKEN` because it doesn't have access right to the version api.

I'm not sure what this means. It sounds like some kind of "bootstrap problem"?

---

I noticed GitLab has a `CI_SERVER_VERSION_MAJOR` predefined variable? Would supporting that be easier than "pinning" yourself to a specific GitLab server version?

Author: We need the server major and minor version to check for specific features and i don't like to pass multiple variables.
This shoul dbe reusable for other platforms later if required.

if you like to use `CI_JOB_TOKEN` for authentication against gitlab, you don't have access to `api/v4/user` and `api/v4/version`. So thats why i like to override it, so we don't need to query the api. The user endpoint is already skiped, when you manually define `gitAuthor`, so only the `version` api request is left. I don't like to add an official option for it yet, maybe later, if this is really required
